app/models/product_importer.rb

```diff
@@ -0,0 +1,461 @@
+      end
+    end
+    @supplier_products
+  end
+
+  def valid_count
+    @valid_entries.count
+  end
+
+  def invalid_count
+    @invalid_entries.count
+  end
+
+  def products_create_count
+    @products_to_create.count + @variants_to_create.count
+  end
+
+  def products_update_count
+    @variants_to_update.count
+  end
+
+  def suppliers_index
+    @suppliers_index || get_suppliers_index
+  end
+
+  def invalid_entries
+    entries = {}
+    @invalid_entries.each do |line_number, data|
+      entries[line_number] = {entry: data[:entry].except(*@non_display_attributes), errors: data[:errors]}
+    end
+    entries
+  end
+
+  def products_to_create
+    entries = {}
+    @products_to_create.merge(@variants_to_create).each do |line_number, data|
+      entries[line_number] = {entry: data[:entry].except(*@non_display_attributes)}
+    end
+    entries
+  end
+
+  def products_to_update
+    entries = {}
+    @variants_to_update.each do |line_number, data|
+      entries[line_number] = {entry: data[:entry].except(*@non_display_attributes)}
+    end
+    entries
+  end
+
+  def products_created_count
+    @products_created + @variants_created
+  end
+
+  def products_updated_count
+    @variants_updated
+  end
+
+  def products_reset_count
+    @products_reset_count || 0
+  end
+
+  def total_saved_count
+    @products_created + @variants_created + @variants_updated
+  end
+
+  def save_all
+    save_all_valid
+    delete_uploaded_file
+  end
+
+  private
+
+  def open_spreadsheet
+    if accepted_mimetype
+      Roo::Spreadsheet.open(@file, extension: accepted_mimetype)
+    else
+      self.errors.add(:importer, 'could not process file: invalid filetype')
+      delete_uploaded_file
+      nil
+    end
+  end
+
+  def accepted_mimetype
+    File.extname(@file.path).in?('.csv', '.xls', '.xlsx', '.ods') ? @file.path.split('.').last.to_sym : false
+  end
+
+  def headers
+    @sheet.row(1)
+  end
+
+  def rows
+    return [] unless @sheet and @sheet.last_row
+    (2..@sheet.last_row).map do |i|
+      @sheet.row(i)
+    end
+  end
+
+  def entries
+    rows.map do |row|
+      Hash[[headers, row].transpose]
+    end
+  end
+
+  def validate_all
+    entries.each_with_index do |entry, i|
+      line_number = i+2 # Roo counts "line 2" as the first line of data
+
+      supplier_validation(line_number, entry)
+      category_validation(line_number, entry)
+      set_update_status(line_number, entry)
+
+      mark_as_valid(line_number, entry) unless entry_invalid?(line_number)
+    end
+
+    count_existing_products
+
+    delete_uploaded_file if item_count.zero? or valid_count.zero?
+  end
+
+  def count_existing_products
+    @suppliers_index.each do |supplier_name, supplier_id|
+      if supplier_id
+        products_count = Spree::Variant.joins(:product).
+          where('spree_products.supplier_id IN (?)
+          AND spree_variants.is_master = false
+          AND spree_variants.deleted_at IS NULL', supplier_id).
+          count
```

Comment:
Well, actually since `products_for_supplier()` returns an `ActiveRecord::Relation` it will hit the DB only when finally invoked with `.count`. Making only the `COUNT(*)` query.

What can be improved is to make the query only once, I need to check it in console but the `||=` memoization should get you covered.